rtcpeerconnection.go

```diff
@@ -233,6 +225,67 @@ func (pc *RTCPeerConnection) initConfiguration(configuration RTCConfiguration) e
 	return nil
 }
 
+// OnDataChannel sets an event handler which is invoked when a data
+// channel message arrives from a remote peer.
+func (pc *RTCPeerConnection) OnDataChannel(f func(*RTCDataChannel)) {
+	pc.Lock()
+	defer pc.Unlock()
+	pc.onDataChannelFn = f
+}
+
+func (pc *RTCPeerConnection) onDataChannel(dc *RTCDataChannel) {
+	pc.RLock()
+	defer pc.RUnlock()
+	if pc.onDataChannelFn == nil || dc == nil {
+		return
+	}
+
+	// Run this synchronously to allow setup done in onDataChannelFn()
```

Comment:
In the past doing synchronous callbacks has caused problems by causing a deadlock when calling methods on the peer connection in the callback. E.g.: the peer connection lock is already locked by onDataChannel, any method called in the callback func that also tries to acquire the lock can cause a deadlock.
Maybe it is sufficient to run both the ``onDataChannelFn`` and ``onOpen`` after each other but in a new routine? This way they are called in the right order but still asynchronous. Alternatively, I think we need to at least release the lock when calling the callback funcs.

Reply:
Yeah, doing it this way didn't give me a great feeling... The problem with doing it async though is that there's a race between kicking off the supplied onDataChannelFn and receiving the next DataChannelEvent in dataChannelEventHandler() -- if that event is a message, and the OnMessage handler hasn't been set yet, the message is missed. I'll think about it some more, because this does feel awfully kludgy.

Reply:
I'm entirely willing to believe that I'm thinking about this the wrong way... But I'm having a hard time figuring out how to achieve my goal of ensuring that data channel operations are serialized, at least for setup.

The code I'm trying to support looks like this:
```go
	pc.OnDataChannel(func(dc *webrtc.RTCDataChannel) {
		logs.Infof("New data channel %s-%d", dc.Label, dc.ID)
		dc.OnOpen(func() {
			logs.Infof("Opened data channel %s-%d", dc.Label, dc.ID)
		})

		dc.OnMessage(func(payload datachannel.Payload) {
			logs.Debugf("Got datachannel message: %#v", payload)
			switch p := payload.(type) {
			case *datachannel.PayloadBinary:
				var msg ChannelMessage
				if err := json.Unmarshal(p.Data, &msg); err != nil {
					logs.Errorf("Failed to unmarshal %s into ChannelMessage: %s", p.Data, err)
				}
				prx.handleChannelMessage(client, pc, dc, &msg)
			case *datachannel.PayloadString:
				logs.Noticef("Got string datachannel message: %s", p.Data)
			default:
				logs.Errorf("Unhandled datachannel message: %s-%s", dc.Label, p.PayloadType())
			}
		})
	})
```

So the issue is that without the enforced serialization it's possible (likely, in my testing) that a data channel message comes in before OnMessage is set.

Another kludge that occurs to me is to do something like this:
```go
func (pc *RTCPeerConnection) onDataChannel(dc *RTCDataChannel) chan struct{} {
	pc.RLock()
	defer pc.RUnlock()
	done := make(chan struct{})
	if pc.onDataChannelHandler == nil || dc == nil {
                close(done)
		return done
	}

	// Run this synchronously to allow setup done in onDataChannelFn()
	// to complete before datachannel event handlers might be called.
	go func(f func(*RTCDataChannel), dc *RTCDataChannel) {
		f(dc)
		close(done)
		dc.onOpen()
	}(pc.onDataChannelHandler, dc)

	return done
}
```

And then callers which care about synchronization (only dataChannelEventHandler() for now) can wait for the channel to close before proceeding.

```go
func (pc *RTCPeerConnection) dataChannelEventHandler(e network.DataChannelEvent) {
	switch event := e.(type) {
	case *network.DataChannelCreated:
		id := event.StreamIdentifier()
		newDataChannel := &RTCDataChannel{ID: &id, Label: event.Label, rtcPeerConnection: pc, ReadyState: RTCDataChannelStateOpen}
		pc.Lock()
		pc.dataChannels[e.StreamIdentifier()] = newDataChannel
		pc.Unlock()

		<-pc.onDataChannel(newDataChannel)
...
}
```

Reply:
I'm gonna run with this idea for now... You guys let me know if it's hideous or if better alternatives present themselves to you.

Reply:
After some discussion in the slack channel, I decided to update all of the on* wrappers to return a done channel which callers can check if they care about waiting. This is internal-only, not exposed via public API. It just gives us more flexibility for handling things behind the scenes.